habitat-lab/habitat/tasks/nav/nav.py

```diff
@@ -1052,20 +1104,57 @@ def update_metric(
 @registry.register_task_action
 class MoveForwardAction(SimulatorTaskAction):
     name: str = "move_forward"
-
-    def step(self, *args: Any, **kwargs: Any):
+    def step(self, task, *args: Any, **kwargs: Any):
         r"""Update ``_metric``, this method is called from ``Env`` on each
         ``step``.
         """
+        from habitat.tasks.rearrange.rearrange_sim import RearrangeSim
+        if type(self._sim) == RearrangeSim:
+            actuation = self._sim.config.agents[0].action_space[1].actuation.amount
+            trans = self._sim.robot.base_transformation
+            local_pos = np.array([actuation, 0, 0])
+            global_pos = trans.transform_point(local_pos)
+            snapped_global_pos = self._sim.step_filter(
+                trans.translation, global_pos
+            )
+            if snapped_global_pos == global_pos:
+                self._sim.robot.base_pos = snapped_global_pos
+                task._is_navmesh_violated = False
+            else:
+                task._is_navmesh_violated = True
         return self._sim.step(HabitatSimActions.move_forward)
 
 
+def rotate_action(sim, direction: str):
```

Comment:
This function has lots of errors in it. Pasting the function after debugging:

def rotate_action(sim, direction: str, **kwargs: Any):
    from habitat.tasks.rearrange.rearrange_sim import RearrangeSim

    assert direction in ["left", "right"]
    if isinstance(sim, RearrangeSim):
        actuation = sim.config.agents[0].action_space[2].actuation.amount
        if "articulated_agent_start_angle" not in dir(sim):
            sim.articulated_agent_start_angle = kwargs[
                "task"
            ]._nav_to_info.articulated_agent_start_angle
            sim.current_angle = sim.articulated_agent_start_angle

        if direction == "right":
            actuation = -1 * actuation

        sim.updated_angle = (
            sim.current_angle + actuation * np.pi / 180
        )
        sim.articulated_agent.base_rot = sim.updated_angle
        sim.current_angle = sim.updated_angle